front/components/assistant_builder/ActionScreen.tsx

```diff
@@ -181,6 +182,38 @@ export default function ActionScreen({
     }
   };
 
+  const [retrievalValid, setRetrievalValid] = useState(true);
+  const [processValid, setProcessValid] = useState(true);
+  const [dustAppRunValid, setDustAppRunValid] = useState(true);
+  const [tablesQueryValid, setTablesQueryValid] = useState(true);
+
+  useEffect(() => {
```

Comment:
Feels kind of weird to use an effect that setStates on something that really isn't state and is 100% derived from builder state no ? I think this is much less efficient

Reply:
Well wrapped in useEffect there is no perf issue here no?

The main driver here is to remove all instances of generic files with action specific code. Does that make sense as a trade off and are you ok directionally with the final proposal for that particular use case (validation)?

Reply:
I don't think there should be any state or use effects here, as that triggers re-renders even though it's not needed (since you can compute action config validity using state you already have)

Reply:
curious what @flvndvd thinks ? I could be wrong here

Reply:
(Definitely aligned that we should co-locate validation code with actions code, but maybe simple pure functions exported that take a builder state and return a bool is what we really need here ?)

Reply:
Henry has a point, and this method might lead to a few side effects. The `useEffect` hook is here since `setActionsValid` isn't directly passed to the action components. Although this approach reduces props drilling, it can make the code a bit harder to read.
This logic allows the action components to update the parent component (in this case, `ActionScreen`), ensuring it remains "aware" of the validity state.
However, props drilling does have its drawbacks, and that's why we might want to consider an alternative approach using `forwardRef`. This method could offer a more elegant and straightforward solution by exposing a simple method like `isActionValid()`.

Reply:
I think we could even skip a forwardref entirely, and just export a function like "isDustAppRunConfigValid(builderState)" (etc..) and have a "isActionsConfigVal(builderState)" function that imports each action config validator. WDYT ?

Reply:
Yep that could work as well!

Reply:
Happy to refactor.

So, if I understand correctly, the idea is to package these in exposed methods on each Action + one that coalesce them in ActionScreen and call that method from the Assistant Builder's useEffect.

Reply:
Way better 🙌